### PR TITLE
Fix `Bastion` deletion when `SSHAccess` is disabled

### PR DESCRIPTION
**Pull request by AleksandarSavchev**
**How to categorize this PR?**

/area security
/kind bug

**What this PR does / why we need it**:
This PR fixes a bug which was causing `Bastion` resources in the garden cluster to not be deleted when the Shoots `SSHAccess` is disabled. When disabling `SSHAccess` we delete all `Bastion` resources in the shoot namespace on the seed and wait for their deletion. The `Bastion` resources in the garden cluster are marked for deletion and have a finalizer which cannot be removed since the `Bastion` admission plugin currently does not allow mutate operation on `Bastion` resource when `SSHAccess` is disabled on the Shoot cluster.

This PR fixes this issue by allowing mutate operation on `Bastion` resources when they have `DeletionTimestamp` set.
 
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```bugfix operator
A bug was fixed which was causing existing `Bastion` resources on the garden cluster to not be deleted when `SSHAccess` is disabled on a Shoot cluster.
```

### CODE REVIEW

File: plugin/pkg/bastion/validator/admission.go [hunk 1]
```diff
@@ -150,7 +150,7 @@ func (v *Bastion) Admit(ctx context.Context, a admission.Attributes, _ admission
 	}
 
 	// ensure shoot SSH access is not disabled
-	if shoot.Spec.Provider.WorkersSettings != nil && shoot.Spec.Provider.WorkersSettings.SSHAccess != nil && !shoot.Spec.Provider.WorkersSettings.SSHAccess.Enabled {
+	if shoot.Spec.Provider.WorkersSettings != nil && shoot.Spec.Provider.WorkersSettings.SSHAccess != nil && !shoot.Spec.Provider.WorkersSettings.SSHAccess.Enabled && bastion.DeletionTimestamp == nil {
```
Comment (ialidzhikov) [inline plugin/pkg/bastion/validator/admission.go]: Not strictly related to this PR but there is already a helper func that can be reused to simplify the check:
```suggestion
	if bastion.DeletionTimestamp == nil && !helper.ShootEnablesSSHAccess(shoot) {
```

File: plugin/pkg/bastion/validator/admission_test.go [hunk 2]
```diff
@@ -264,6 +264,25 @@ var _ = Describe("Bastion", func() {
...
+			now := metav1.Now()
+			bastion.DeletionTimestamp = &now
+
+			coreClient.AddReactor("get", "shoots", func(action testing.Action) (bool, runtime.Object, error) {
+				return true, shoot, nil
+			})
+
+			oldBastion := bastion.DeepCopy()
+			oldBastion.ObjectMeta.Finalizers = []string{"foo"}
+			bastion.ObjectMeta.Finalizers = []string{""}
```
Comment (ialidzhikov) [inline plugin/pkg/bastion/validator/admission_test.go]: Nit: A more real-world scenario is
```suggestion
			bastion.ObjectMeta.Finalizers = []string{}
```

or

```suggestion
			bastion.ObjectMeta.Finalizers = nil
```

I guess no one sets a finalizer that is empty string. Most probably it is even forbidden by the API.

**Review by ialidzhikov — COMMENTED** [review]
/lgtm

File: plugin/pkg/bastion/validator/admission_test.go [hunk 3]
```diff
@@ -71,6 +72,11 @@ var _ = Describe("Bastion", func() {
 					SeedName: pointer.String(seedName),
 					Provider: gardencore.Provider{
 						Type: provider,
+						Workers: []gardencore.Worker{
+							{
+								Name: workerName,
```
Comment (plkokanov) [inline plugin/pkg/bastion/validator/admission_test.go:77]: Why is this necessary now?

**Review by plkokanov — COMMENTED** [review]
Just a quick question, otherwise lgtm.

File: plugin/pkg/bastion/validator/admission_test.go (hunk 3 above)
Comment (AleksandarSavchev) [reply plugin/pkg/bastion/validator/admission_test.go:77]: We are now using [ShootEnablesSSHAccess()](https://github.com/gardener/gardener/blob/54b0182e8f7231ca7b5de88bf179fb6dc4192851/pkg/apis/core/helper/helpers.go#L501-L505) which returns `false` when `SSHAccess` is disabled or the `Shoot` is workerless. When the `Shoot` is workerless we accept that it has disabled `ssh` and forbid mutate operations on `Bastion` resources, hence we need to add a worker to the test `Shoot`.
